web-api/src/middleware/request_context.rs

```diff
@@ -0,0 +1,184 @@
+    type Error = anyhow::Error;
+
+    fn try_from(value: &HeaderValue) -> Result<Self, Self::Error> {
+        let header = value.to_str().map_err(|_| {
+            anyhow!(
+                "Non Utf-8 compatible {TENANT_ID_HEADER} header: {:?}",
+                String::from_utf8_lossy(value.as_bytes())
+            )
+        })?;
+        header
+            .parse::<Uuid>()
+            .map_err(|_| anyhow!("Non UUID {TENANT_ID_HEADER} header: {header}"))
+            .map(TenantId)
+    }
+}
+
+#[derive(Clone, Copy, Debug, derive_more::Display, Serialize)]
+#[serde(transparent)]
+pub(crate) struct RequestId(Uuid);
+
+impl RequestId {
+    pub(crate) fn generate() -> Self {
+        Self(Uuid::new_v4())
+    }
+
+    pub(crate) const fn missing() -> Self {
+        Self(Uuid::nil())
+    }
+}
+
+/// Setups the call context.
+///
+/// This make the `RequestId` and `TenantId` available as extensions and sets up tracing for all calls.
+///
+/// The `TenantId` is seen as _required_, i.e. the middleware will fail any
+/// request where it can't find a tenant id.
+pub(crate) fn setup_request_context<A, S>(
+    request: ServiceRequest,
+    service: &S,
+) -> impl Future<Output = Result<ServiceResponse<BoxBody>, actix_web::Error>> + 'static
+where
+    A: AsRef<tenants::Config> + 'static,
+    S: Service<ServiceRequest, Response = ServiceResponse<BoxBody>, Error = actix_web::Error>,
+    S::Future: 'static,
+{
+    let request_id = RequestId::generate();
+    let config = request.app_data::<Data<A>>().unwrap().get_ref().as_ref();
+
+    let tenant_id = match extract_tenant_id(config, &request) {
+        Ok(id) => id,
+        Err(error) => {
+            let response = middleware_failure(
+                "setup_request_context",
+                request,
+                Some(request_id),
+                None,
+                error,
+            );
+            return Either::Left(future::ok(response));
+        }
+    };
+
+    let span = info_span!(
+        "request",
+        path = %request.request().path(),
+        method = %request.request().method(),
+        request_id = %request_id,
+        tenant_id = %tenant_id,
+    );
+
+    trace!(parent: &span, "request received");
+
+    let context = Arc::new(RequestContext {
+        tenant_id,
+        request_id,
+    });
+
+    request.extensions_mut().insert(context);
+
+    Either::Right(
+        service
+            .call(request)
+            .instrument(span.clone())
+            .inspect(|_| trace!(parent: span, "request processed")),
+    )
+}
+
+const TENANT_ID_HEADER: &str = "X-Tenant-Id";
+
+fn extract_tenant_id(
+    config: &tenants::Config,
+    request: &ServiceRequest,
+) -> Result<TenantId, anyhow::Error> {
+    let header_value = request
+        .headers()
+        .get(TENANT_ID_HEADER)
+        .map(TenantId::try_from)
+        .transpose()?;
+
+    match header_value {
+        //FIXME in follow up PR this ID will be fetched from the database
```

Comment:
> should continue to work as is 

It does.

Like written it will fetch the id for the legacy tenant form the database.  The first time the new rust code is started  (of the follow up PR) it will randomly generate the id and insert it there (and rename the `public` schema to `t:{legacy_tenant_id}`).

This approach is easier to implement then any of the alternatives AFIK, as it makes the whole thing about having different `modes` and other special cases through the code unnecessary and renaming a schema is pretty trivial (one line of sql).

It also has the neat side effect of making migrating legacy tenants to non-legacy single tenants or multi tenants much easier or allow us to add additional tenants on a single tenant mode deployment. Non of this played a role for why I choose to implement it this way, but they are nice anyway.

The reason I didn't purse this approach from the beginning was that I though  there would be certain problems with this which turned out to be non-issues.

Reply:
This is more complicated simply because we are doing work that is not needed. Generating a random id here is logically wrong because the id must come from the control panel and the infrastructure not our code.

> It also has the neat side effect of making migrating legacy tenants to non-legacy single tenants or multi tenants much easier or allow us to add additional tenants on a single tenant mode deployment

The only use case we will have, much later in the future, will be to move the current tenants onto the multi-tenant infrastructure. Till then we will have two separate infrastructures and we might deploy customers on one or the other depending on their needs.